### ISSUE TITLE
cannot open output file ...\libxamarin-app.so: function not supported

### ISSUE THREAD

**Issue by YouKnowThem** [xamarin/xamarin-android#7888]
### Android application type

.NET Android (net7.0-android, etc.)

### Affected platform version

VS 2022 17.5.1 and 17.6.0 Preview 2.0, Windows 11 and Windows 10, .NET 7, Microsoft.Andriod.Sdk.Windows 33.0.26

### Description

Unable to build the android output of the default .NET MAUI (xaml) project (no code modifications) with the error 
"C:\Program Files\dotnet\packs\Microsoft.Android.Sdk.Windows\33.0.26\tools\Xamarin.Android.Common.targets(2002,3): error XA3007: stderr | ld: error: cannot open output file obj\Debug\net7.0-android\app_shared_libraries\x86_64\libxamarin-app.so: function not supported"

When switching the build output in VS: Tools-->Options-->Projects and Solutions-->Build And Run-->MSBuild project build output verbosity to Diagnostics I have found that the native linker was not able to create the file **libxamarin-app.so**. When running the command manually with the **--verbose** attribute the output was:
```
L:\MauiApp1\MauiApp1>"C:\Program Files\dotnet\packs\Microsoft.Android.Sdk.Windows\33.0.26\tools\binutils\bin\ld.EXE" --unresolved-symbols=ignore-in-shared-libs --export-dynamic -soname libxamarin-app.so -z relro -z noexecstack --enable-new-dtags --eh-frame-hdr -shared --build-id --warn-shared-textrel --fatal-warnings -m elf_x86_64 obj\Debug\net7.0-android\android\typemaps.x86_64.o obj\Debug\net7.0-android\android\environment.x86_64.o obj\Debug\net7.0-android\android\compressed_assemblies.x86_64.o obj\Debug\net7.0-android\android\jni_remap.x86_64.o -o obj\Debug\net7.0-android\app_shared_libraries\x86_64\libxamarin-app.so --verbose
ld: obj\Debug\net7.0-android\android\typemaps.x86_64.o
ld: obj\Debug\net7.0-android\android\environment.x86_64.o
ld: obj\Debug\net7.0-android\android\compressed_assemblies.x86_64.o
ld: obj\Debug\net7.0-android\android\jni_remap.x86_64.o
ld: error: cannot open output file obj\Debug\net7.0-android\app_shared_libraries\x86_64\libxamarin-app.so: function not supported
```
You might already spot the issue, and you are correct. It is because the linker and the project are **not on the same drive**.
L is a virtual RAM drive. When creating the project on the C drive, there is no issue and the build works fine.

### Steps to Reproduce

1. On Windows 11 or 10 install VS 2022 with .NET MAUI worksets
2. Get ImDisk or something else to create a virtual RAM disk (at least 1 GB in size), NTFS
3. Create a new .NET MAUI project in VS 2022 inside the RAM disk
4. Build the project against Andriod Emulator (I used the default Pixel 5 API 33)
5. The error comes up

### Did you find any workaround?

Create and build on the default C drive instead of a different drive (in my example the virtual RAM disk)

**Comment by dellis1972** [comment]
@YouKnowThem 

Windows has a problem moving files from a temp folder which is not on the same drive as the destination file. The `ld` tool probably uses temp files to do its works. 

You might try to alter the `TEMP` environment variable to point to a location on the Ram disk to see if that helps.

**Comment by dellis1972** [comment]
@grendello this might be of interest, any idea if we can override the temp directory when using tools like `ld` so that they don't use the global temp folder? (assuming this issue is caused by that).

**Comment by YouKnowThem** [comment]
I have no idea if it supports your theory or not, but without dealing with the TEMP environment variable, instead of the **libxamarin-app.so** being created in obj/Debug/net7.0-android/app_shared_libraries/x86_64, there is a zero byte size file with a name of **libxamarin-app.so.tmp9c7fd56**. I will fiddle with the TEMP environment variable shortly. Stay tuned.

**Comment by YouKnowThem** [comment]
I had two entries, that looked promising: "TEMP" and "TMP". Each of them in both the user and system wide environment settings. I changed all four of them one after another. Restarted VS and even restarting the machine to be on the safe side. All to no avail, unfortunately.
TLDR: Changing the `TEMP` environment variable to point to the RAM disk did not work. But thanks for the idea.

If anything other comes to mind, I am happy to try it out.

**Comment by grendello** [comment]
@dellis1972 @YouKnowThem  looking at the sources, `TMPDIR` environment variable should do it

**Comment by YouKnowThem** [comment]
@grendello as far as I understand it, `TMPDIR` is a typical linux environment variable. In Microsoft Windows it is `TEMP` or `TMP`. And setting those 2 I have already tried without success. Setting up a new `TMPDIR` variable did also not suffice. The same error is still there.
Apologies if I did not get what you wanted to say.

**Comment by grendello** [comment]
@YouKnowThem `TMPDIR` is specific to Unix, yes. LLVM also checks `TEMPDIR`, `TMP` and `TEMP` on Unix. For Windows, [they check](https://github.com/llvm/llvm-project/blob/llvmorg-15.0.7/llvm/lib/Support/Windows/Path.inc#L1451-L1496) for `TMP`, `TEMP` and `USERPROFILE` (which almost mirrors what [Win32 API](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-gettemppatha#remarks) does, with the exception that LLVM falls back to `c:\\Temp` instead of to Windows directory). If neither one of those work, I'm not sure how else to override the temporary directory location.
